query-engine/driver-adapters/src/proxy.rs

```diff
@@ -173,169 +177,211 @@ pub struct Query {
     pub args: Vec<JSArg>,
 }
 
+fn conversion_error(args: &std::fmt::Arguments) -> QuaintError {
+    let msg = match args.as_str() {
+        Some(s) => Cow::Borrowed(s),
+        None => Cow::Owned(args.to_string()),
+    };
+    QuaintError::builder(ErrorKind::ConversionError(msg)).build()
+}
+
+macro_rules! conversion_error {
+    ($($arg:tt)*) => {
+        conversion_error(&format_args!($($arg)*))
+    };
+}
+
 /// Handle data-type conversion from a JSON value to a Quaint value.
 /// This is used for most data types, except those that require connector-specific handling, e.g., `ColumnType::Boolean`.
 fn js_value_to_quaint(
     json_value: serde_json::Value,
     column_type: ColumnType,
     column_name: &str,
-) -> QuaintValue<'static> {
+) -> quaint::Result<QuaintValue<'static>> {
     //  Note for the future: it may be worth revisiting how much bloat so many panics with different static
     // strings add to the compiled artefact, and in case we should come up with a restricted set of panic
     // messages, or even find a way of removing them altogether.
     match column_type {
         ColumnType::Int32 => match json_value {
             serde_json::Value::Number(n) => {
                 // n.as_i32() is not implemented, so we need to downcast from i64 instead
-                QuaintValue::int32(n.as_i64().expect("number must be an i32") as i32)
-            }
-            serde_json::Value::String(s) => {
-                let n = s.parse::<i32>().expect("string-encoded number must be an i32");
-                QuaintValue::int32(n)
+                n.as_i64()
+                    .ok_or(conversion_error!("number must be an integer"))
+                    .and_then(|n| -> quaint::Result<i32> {
+                        n.try_into()
```

Comment:
This one changes the behavior a little: previously we would silently overflow/undeflow here, and now it's an error if it doesn't fit into i32.